model/convert/service_event.go

```diff
@@ -0,0 +1,326 @@
+// state. This acts as the conversion from the Cadence type to the flow-go type.
+func ServiceEvent(event flow.Event) (*flow.ServiceEvent, error) {
+
+	// depending on type of Epoch event construct Go type
+	switch event.Type {
+	case flow.EventEpochSetup:
+		return convertServiceEventEpochSetup(event)
+	case flow.EventEpochCommit:
+		return convertServiceEventEpochCommit(event)
+	default:
+		return nil, fmt.Errorf("invalid event type: %s", event.Type)
+	}
+}
+
+// convertServiceEventEpochSetup converts a service event encoded as the generic
+// flow.Event type to a ServiceEvent type for an EpochSetup event
+func convertServiceEventEpochSetup(event flow.Event) (*flow.ServiceEvent, error) {
+
+	// decode bytes using jsoncdc
+	payload, err := json.Decode(event.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal event payload: %w", err)
+	}
+
+	// parse cadence types to required fields
+	setup := new(flow.EpochSetup)
+
+	// NOTE: variable names prefixed with cdc represent cadence types
+	cdcEvent := payload.(cadence.Event)
+
+	// extract simple fields
+	setup.Counter = uint64(cdcEvent.Fields[0].(cadence.UInt64))
+	setup.FirstView = uint64(cdcEvent.Fields[2].(cadence.UInt64))
+	setup.FinalView = uint64(cdcEvent.Fields[3].(cadence.UInt64))
+	randomSrcHex := string(cdcEvent.Fields[5].(cadence.String))
+	setup.RandomSource, err = hex.DecodeString(randomSrcHex)
+	if err != nil {
+		return nil, fmt.Errorf("could not decode random source hex: %w", err)
+	}
+	setup.DKGPhase1FinalView = uint64(cdcEvent.Fields[6].(cadence.UInt64))
+	setup.DKGPhase2FinalView = uint64(cdcEvent.Fields[7].(cadence.UInt64))
+	setup.DKGPhase3FinalView = uint64(cdcEvent.Fields[8].(cadence.UInt64))
+
+	// parse cluster assignments
+	cdcClusters := cdcEvent.Fields[4].(cadence.Array).Values
+	assignments, err := convertClusterAssignments(cdcClusters)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert cluster assignments: %w", err)
+	}
+	setup.Assignments = assignments
+
+	// parse epoch participants
+	cdcParticipants := cdcEvent.Fields[1].(cadence.Array).Values
+	setup.Participants, err = convertParticipants(cdcParticipants)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert participants: %w", err)
+	}
+
+	// construct the service event
+	serviceEvent := &flow.ServiceEvent{
+		Type:  flow.ServiceEventSetup,
+		Event: setup,
+	}
+
+	return serviceEvent, nil
+}
+
+// convertServiceEventEpochCommit converts a service event encoded as the generic
+// flow.Event type to a ServiceEvent type for an EpochCommit event
+func convertServiceEventEpochCommit(event flow.Event) (*flow.ServiceEvent, error) {
+
+	// decode bytes using jsoncdc
+	payload, err := json.Decode(event.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("could not unmarshal event payload: %w", err)
+	}
+
+	// parse cadence types to Go types
+	commit := new(flow.EpochCommit)
+	commit.Counter = uint64(payload.(cadence.Event).Fields[0].(cadence.UInt64))
```

Comment:
Same here, maybe check the fields' length before accessing them, and conditionally cast, i.e. handle casting failures gracefully